target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c

```diff
@@ -979,37 +1101,93 @@ static void rtl93xx_phylink_mac_link_up(struct dsa_switch *ds, int port,
 static void rtl83xx_get_strings(struct dsa_switch *ds,
 				int port, u32 stringset, u8 *data)
 {
+	const struct rtl83xx_mib_desc *mib;
+	struct rtl838x_switch_priv *priv = ds->priv;
+	int mib_size;
+
 	if (stringset != ETH_SS_STATS)
 		return;
 
-	for (int i = 0; i < ARRAY_SIZE(rtl83xx_mib); i++)
-		ethtool_puts(&data, rtl83xx_mib[i].name);
+	switch(priv->family_id) {
+	case RTL9300_FAMILY_ID:
+		mib = rtl930x_mib;
+		mib_size = ARRAY_SIZE(rtl930x_mib);
+		break;
+	case RTL9310_FAMILY_ID:
+		mib = rtl931x_mib;
+		mib_size = ARRAY_SIZE(rtl931x_mib);
+		break;
+	default:
+		mib = rtl83xx_mib;
+		mib_size = ARRAY_SIZE(rtl83xx_mib);
+		break;
+	}
+
+	for (int i = 0; i < mib_size; i++, mib++)
+		ethtool_puts(&data, mib->name);
+}
+
+static uint64_t mib_read_common(struct rtl838x_switch_priv *priv, int port, int mib_size, int mib_offset)
+{
+	uint64_t ret = 0;
+
+	if (mib_size == 2) {
+		ret = sw_r32(priv->r->stat_port_std_mib + (port << 8) + 248 - mib_offset);
+		ret <<= 32;
+	}
+
+	ret |= sw_r32(priv->r->stat_port_std_mib + (port << 8) + 252 - mib_offset);
+
+	return ret;
 }
 
 static void rtl83xx_get_ethtool_stats(struct dsa_switch *ds, int port,
 				      uint64_t *data)
 {
 	struct rtl838x_switch_priv *priv = ds->priv;
 	const struct rtl83xx_mib_desc *mib;
-	u64 h;
+	int mib_size;
 
-	for (int i = 0; i < ARRAY_SIZE(rtl83xx_mib); i++) {
-		mib = &rtl83xx_mib[i];
+	switch (priv->family_id) {
+	case RTL9300_FAMILY_ID:
+		mib = rtl930x_mib;
+		mib_size = ARRAY_SIZE(rtl930x_mib);
+		break;
+	case RTL9310_FAMILY_ID:
+		mib = rtl931x_mib;
+		mib_size = ARRAY_SIZE(rtl931x_mib);
+		break;
+	default:
+		mib = rtl83xx_mib;
+		mib_size = ARRAY_SIZE(rtl83xx_mib);
+		break;
+	}
 
-		data[i] = sw_r32(priv->r->stat_port_std_mib + (port << 8) + 252 - mib->offset);
-		if (mib->size == 2) {
-			h = sw_r32(priv->r->stat_port_std_mib + (port << 8) + 248 - mib->offset);
-			data[i] |= h << 32;
+	if (priv->r->mib_read) {
+		for (int i = 0; i < mib_size; i++, mib++)
+			data[i] = priv->r->mib_read(port, mib->size, mib->offset);
```

Comment:
It is odd to have an iterator i and then not using it.

`mib[i].name, mib[i].offset`